DRIVERS-2035 Use minimum RTT for CSOT maxTimeMS calculation instead of 90th percentile

ShaneHarvey opened the pull request:
https://jira.mongodb.org/browse/DRIVERS-2035

Please complete the following before merging:

- [X] Update changelog.
- [X] Make sure there are generated JSON files from the YAML test files.
- [x] Test changes in at least one language driver: https://spruce.mongodb.com/version/63e568b07742ae5e72e81ce0/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC
- [x] Test these changes against all server versions and topologies (including standalone, replica set, sharded clusters, and serverless).

benjirewis commented on source/client-side-operations-timeout/client-side-operations-timeout.rst:
```diff
@@ -899,6 +899,15 @@ introduce a new knob and increase the API surface of drivers without providing
+A previous version of this spec used the 90th percentile RTT to short
+circuit operations that might otherwise fail with a socket timeout.
+We decided to change this logic to avoid canceling operations that may
+have a high chance of succeeding and also removes a dependency on t-digest.
```
```suggestion
have a high chance of succeeding; this change also removes a dependency on t-digest.
```

matthewdale commented on source/client-side-operations-timeout/tests/command-execution.yml:
```diff
@@ -39,7 +39,7 @@ tests:
               failCommands: ["hello", "isMaster"]
```
The 10 sample minimum is an attempt to reduce the probability that early RTT outliers cause unnecessary operation skipping. Keep in mind that the "min RTT" logic currently in the Go driver was not added as part of CSOT, but was an optimization added with Go driver connection pool improvements. I decided to be conservative with the implementation because it was a novel and untested concept at the time and impacts all Go driver users.

We haven't encountered any known issues with that "min RTT" implementation, so it makes sense to be less conservative with the opt-in CSOT behavior. However, trusting the first RTT sample on startup seems dangerous and may lead to to increased operation errors immediately after connecting, until more RTT samples are collected. I recommend we collect at least 2 samples before assuming it's an accurate representation of the actual RTT.

I don't have any stats on the impact of requiring 1 vs 2 RTT samples, only the observation that [networks are unreliable](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing#The_fallacies) and that any single RTT sample may have a nontrivial probability of being bogus. That may be especially true while an application is starting up.

As far as using a fixed number of samples (e.g. 10), that seems totally reasonable.

ShaneHarvey replied:
Updated the spec to use a min of 2 and max of 10 samples. I also updated the tests to wait (heuristically) for >=2 samples.

ShaneHarvey commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -812,24 +810,25 @@ on a dedicated connection, for example:
         lock = Mutex()
         movingAverage = MovingAverage()
-        rttDigest = TDigest() # for 90th percentile RTT calculation
+        rttMin = MinWindow(max_window_size=10) # for min RTT calculation
```
Fixed. The peusdo-python is now explicit about the behavior using a deque: https://docs.python.org/3/library/collections.html#deque-objects

ShaneHarvey commented:
@matthewdale this is ready for another look. I've updated the spec to require at least 2 RTT samples before enabling the short-circuit logic and updated the tests to account for this.

matthewdale commented on source/server-discovery-and-monitoring/server-monitoring.rst:
```diff
@@ -541,12 +541,12 @@ Clients MUST ignore the response to the hello or legacy hello command when measu
-clients MUST use average and 90th percentile round trip times from the RTT
-task.
+clients MUST set the average and minimum round trip times from the RTT task as the
+"roundTripTime" and "minRoundTripTime" fields, respectively.
```
This requirement has actually lead to significant confusion when implementing features in the Go driver that depend on up-to-date avg/min RTT values. Using values for average and minimum RTT pulled from server descriptions can lead to difficult-to-diagnose bugs unless devs realize that the RTT values on server descriptions are never updated after handshake. The scope for Go Driver 2.0 currently includes removing all values from server descriptions that are not derived directly from the connection/handshake process. See [GODRIVER-2691](https://jira.mongodb.org/browse/GODRIVER-2691) for more details.

Should we remove this requirement and the corresponding `roundTripTime` and `minRoundTripTime` fields from a `ServerDescription`?

P.S. I realize this is only updating the existing requirement, but this could be a good time to reconsider the need for these fields.

ShaneHarvey replied:
I don't agree because even fields that are reported by the server in the hello command response can become stale. For example, hosts, isWritablePrimary, $clusterTime, operationTime, etc.. In fact, almost every single field can become stale right after the server sends it.

matthewdale reviewed (approved):
One question, but otherwise looks good 👍
